Feature/ral 8 sliding window algorithm

ParthZee opened the pull request.

ParthZee committed:
feat: setup basic server responding with 'Hello World'

ParthZee committed:
feat: add basic limited and unlimited api routes

ParthZee committed:
docs: add rate-limiting.md file

ParthZee committed:
feat: add rateLimter middleware to the /limited route

ParthZee committed:
feat: implement IP based in-memory rate limiter

ParthZee committed:
fix: rename rateLimiter.js middleware to fixedWindowRateLimiter.js

ParthZee committed:
feat: configure jest testing framework for CommonJS

ParthZee committed:
feat: configure jest to support ECMAScript Modules (ESM) as used in the project

ParthZee committed:
test: add unit tests for fixed window rate limiter

ParthZee committed:
feature: add token bucket rate limiter as middleware

ParthZee committed:
feat: implement token bucket rate limiter

ParthZee committed:
feat: add unit tests for token bucket rate limiter and change middleware import in index.js file

ParthZee committed:
refactor: separate express app and server into distinct files

ParthZee committed:
fix: fix response to also send status code 200 for successful requests to the route

ParthZee committed:
test: add E2E tests for token bucket rate limiter and enabled trust proxy in app.js

ParthZee committed:
fix: add comment in jest.config.js file

ParthZee committed:
test: add multiple client test to E2E test and uncomment remaining tests for token bucket rate limiter

ParthZee committed:
refactor: rename both rate limiter middlewares

ParthZee committed:
refactor: rename rate limiter imports in test files

ParthZee committed:
refactor: rename all the rate limiter test files

chintal31 commented on src/middleware/fixedWindow.js:
```diff
@@ -0,0 +1,47 @@
+  const ip = rawIp.startsWith("::ffff:") ? rawIp.slice(7) : rawIp; //slicing as we are getting IPv6-mapped IPv4 address if executed on local machine (eg. ::ffff:127.0.0.1)
+  const oneMin = 60000; //60,000 milliseconds
+
+  // Condition to check if the client has not made request to the route before
```
IMO: This comment may be redundant. Ideally, code should be written to be self-explanatory, especially when using clear and meaningful conditions. In such cases, comments might not add value and can be safely omitted.

chintal31 reviewed (commented):
Great work! The code reflects solid practices — consistent use of camelCase, clear and descriptive variable and function names, and a good separation of concerns. Nicely done! 👍🏽

ParthZee committed:
chore: improve formatting of comments

ParthZee committed:
chore: update 429 error message for clarity

chintal31 reviewed (approved):
Looks good. LGTM! 🎉

ParthZee committed:
feat(redis): setup Redis client and add basic string operations

ParthZee committed:
feat: implement sliding window rate limiter with Redis

ParthZee committed:
test: add unit tests with proper ESM mocking using unstable_mockModule

ParthZee committed:
fix: clear Jest mocks and correct typo in test files and edit package.json

ParthZee committed:
test: add E2E tests for sliding window rate limiter

ParthZee committed:
rename: enable token bucket middleware by renaming file

ParthZee committed:
fix: update client.js to pass the base or radix in the parseInt method

ParthZee committed:
feat: add Retry-After header to the response for consistency and standard support

ParthZee committed:
refactor: improve middleware clarity, config safety, and test stability

- Removed unnecessary await from next() calls in middleware

- Replaced magic numbers with named constants for readability

- Added fallback for process.env.REDIS_PORT to prevent Redis misconfig

- Imported and closed Redis client to resolve lingering worker in E2E tests

- Mocked setHeader in slidingWindow.test.js after adding Retry-After header

ParthZee committed:
chore: add server error handler for improved startup resilience

ParthZee committed:
test: improve Redis isolation, memory handling, and middleware efficiency

- Added client.flushdb() in beforeEach to prevent cross-test interference

- Implemented TTL-based cleanup for ipTracker map to prevent memory leaks

- Moved constants outside the middleware to reduce repeated allocation on every request

ParthZee committed:
chore: update comments for clarity

ParthZee committed:
docs: update README with project overview, features, and test instructions

ParthZee committed:
test: add E2E tests for fixed window rate limiter

ParthZee committed:
docs: update README with Fixed Window E2E test coverage

ParthZee committed:
docs: add Middleware Test Isolation section in README file

chintal31 commented on package.json:
```diff
@@ -0,0 +1,22 @@
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
You can remove the `license` field or keep the field empty string, since this is a non-licensed pet project.

chintal31 commented on src/middleware/slidingWindow.js:
```diff
@@ -0,0 +1,66 @@
+    const cutoffTimestamp = currentTimestamp - WINDOW_SIZE;
+
+    // So min:0 to max: start of the window
+    await client.zremrangebyscore(`${PREFIX}:${ip}`, 0, cutoffTimestamp);
```
The `zremrangebyscore` method is a low-level Redis client command. To improve readability and intent clarity, it would be better to wrap this logic inside a descriptive helper method—e.g., client.cleanUpOldEntries('prefix:ip').

Similarly, instead of using `client.zcard` directly, a wrapper like `client.countActiveEntries ` would make the code more self-explanatory.

These changes don’t alter the underlying behavior but significantly enhance code readability and maintainability.

chintal31 reviewed (commented):
Overall, looks good. 👍🏽

ParthZee committed:
refactor: improve Redis abstraction, config cleanup, and documentation

- Improved code readability and encapsulation of low-level Redis operations

- Replaced repeated ${PREFIX}: expressions with reusable key variable

- Removed license field from package.json as this is a personal project

- Added helper.js to README.md project structure section

chintal31 reviewed (approved):
Nice, looks quite good.
LGTM!
